go/cmd/dolt/commands/sqlserver/server.go

```diff
@@ -63,335 +64,470 @@ func Serve(
-		cli.PrintErr(startError)
-		return
+	var mySQLServer *server.Server
+	InitSQLServer := &svcs.Service{
+		Init: func(context.Context) error {
+			var err error
+			v, ok := serverConfig.(validatingServerConfig)
+			if ok && v.goldenMysqlConnectionString() != "" {
+				mySQLServer, err = server.NewValidatingServer(
+					serverConf,
+					sqlEngine.GetUnderlyingEngine(),
+					newSessionBuilder(sqlEngine, serverConfig),
+					listener,
+					v.goldenMysqlConnectionString(),
+				)
+			} else {
+				mySQLServer, err = server.NewServer(
+					serverConf,
+					sqlEngine.GetUnderlyingEngine(),
+					newSessionBuilder(sqlEngine, serverConfig),
+					listener,
+				)
+			}
+			if errors.Is(err, server.UnixSocketInUseError) {
+				lgr.Warn("unix socket set up failed: file already in use: ", serverConf.Socket)
+				err = nil
+			}
+			return err
+		},
 	}
+	controller.Register(InitSQLServer)
 
-	sqlserver.SetRunningServer(mySQLServer, serverLock)
+	LockMultiRepoEnv := &svcs.Service{
+		Init: func(context.Context) error {
+			if ok, f := mrEnv.IsLocked(); ok {
+				return env.ErrActiveServerLock.New(f)
+			}
+			if err := mrEnv.Lock(serverLock); err != nil {
+				return err
+			}
+			return nil
+		},
+		Stop: func() error {
+			if err := mrEnv.Unlock(); err != nil {
+				cli.PrintErr(err)
+			}
+			return nil
+		},
+	}
+	controller.Register(LockMultiRepoEnv)
 
-	ed = mysqlDb.Editor()
-	mysqlDb.AddSuperUser(ed, LocalConnectionUser, "localhost", serverLock.Secret)
-	ed.Close()
-	if ExternalDisableUsers {
-		mysqlDb.SetEnabled(false)
+	InitLockSuperUser := &svcs.Service{
+		Init: func(context.Context) error {
+			mysqlDb := sqlEngine.GetUnderlyingEngine().Analyzer.Catalog.MySQLDb
+			ed := mysqlDb.Editor()
+			mysqlDb.AddSuperUser(ed, LocalConnectionUser, "localhost", serverLock.Secret)
+			ed.Close()
+			return nil
+		},
 	}
+	controller.Register(InitLockSuperUser)
 
-	var metSrv *http.Server
-	if serverConfig.MetricsHost() != "" && serverConfig.MetricsPort() > 0 {
-		mux := http.NewServeMux()
-		mux.Handle("/metrics", promhttp.Handler())
+	DisableMySQLDbIfRequired := &svcs.Service{
+		Init: func(context.Context) error {
+			if ExternalDisableUsers {
+				mysqlDb := sqlEngine.GetUnderlyingEngine().Analyzer.Catalog.MySQLDb
+				mysqlDb.SetEnabled(false)
+			}
+			return nil
+		},
+	}
+	controller.Register(DisableMySQLDbIfRequired)
 
-		metSrv = &http.Server{
-			Addr:    fmt.Sprintf("%s:%d", serverConfig.MetricsHost(), serverConfig.MetricsPort()),
-			Handler: mux,
-		}
+	var metSrv *http.Server
+	RunMetricsServer := &svcs.Service{
+		Run: func(context.Context) {
+			if serverConfig.MetricsHost() != "" && serverConfig.MetricsPort() > 0 {
+				mux := http.NewServeMux()
+				mux.Handle("/metrics", promhttp.Handler())
+
+				metSrv = &http.Server{
+					Addr:    fmt.Sprintf("%s:%d", serverConfig.MetricsHost(), serverConfig.MetricsPort()),
+					Handler: mux,
+				}
 
-		go func() {
-			_ = metSrv.ListenAndServe()
-		}()
+				_ = metSrv.ListenAndServe()
```

Comment:
TODO: Dropped error here if Listen fails. Might be better to separate Listen from Serve and Listen in Init.